pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,485 @@
+	c.enqueueExternalIPPools(affectedEIPs)
+	klog.V(2).InfoS("Processed Node DELETE event", "nodeName", node.Name, "affectedExternalIPPoolNum", affectedEIPs.Len())
+}
+
+func (c *Cluster) handleUpdateNode(oldObj, newObj interface{}) {
+	node := newObj.(*corev1.Node)
+	oldNode := oldObj.(*corev1.Node)
+	if reflect.DeepEqual(node.GetLabels(), oldNode.GetLabels()) {
+		klog.V(2).InfoS("Processing Node UPDATE event error, labels not changed", "nodeName", node.Name)
+		return
+	}
+	oldMatches, newMatches := c.filterEIPsFromNodeLabels(oldNode), c.filterEIPsFromNodeLabels(node)
+	if oldMatches.Equal(newMatches) {
+		klog.V(2).InfoS("Processing Node UPDATE event error, Node cluster status not changed", "nodeName", node.Name)
+		return
+	}
+	affectedEIPs := oldMatches.Union(newMatches)
+	c.enqueueExternalIPPools(affectedEIPs)
+	klog.V(2).InfoS("Processed Node UPDATE event", "nodeName", node.Name, "affectedExternalIPPoolNum", affectedEIPs.Len())
+}
+
+func (c *Cluster) enqueueExternalIPPools(eips sets.String) {
+	for eip := range eips {
+		c.queue.Add(eip)
+	}
+}
+
+func (c *Cluster) enqueueExternalIPPool(obj interface{}) {
+	eip, ok := obj.(*v1alpha2.ExternalIPPool)
+	if !ok {
+		deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.ErrorS(errDecodingObject, "Processing ExternalIPPool DELETE event error", "obj", obj)
+			return
+		}
+		eip, ok = deletedState.Obj.(*v1alpha2.ExternalIPPool)
+		if !ok {
+			klog.ErrorS(errDecodingObjectTombstone, "Processing ExternalIPPool DELETE event error", "obj", deletedState.Obj)
+			return
+		}
+	}
+	c.queue.Add(eip.Name)
+}
+
+// newClusterMember get Node ip and return a cluster member(ip:clusterMemberlistPort) for joining.
+func (c *Cluster) newClusterMember(node *corev1.Node) (string, error) {
+	nodeAddr, err := k8s.GetNodeAddr(node)
+	if err != nil {
+		return "", fmt.Errorf("obtain IP address from K8s Node failed: %v", err)
+	}
+	member := fmt.Sprintf("%s:%d", nodeAddr, c.bindPort)
+	return member, nil
+}
+
+func (c *Cluster) allClusterMembers() (clusterNodes []string, err error) {
+	nodes, err := c.nodeLister.List(labels.Everything())
+	if err != nil {
+		return nil, fmt.Errorf("listing Nodes error: %v", err)
+	}
+
+	for _, node := range nodes {
+		member, err := c.newClusterMember(node)
+		if err != nil {
+			klog.ErrorS(err, "Get Node failed")
+			continue
+		}
+		clusterNodes = append(clusterNodes, member)
+	}
+	return
+}
+
+func (c *Cluster) filterEIPsFromNodeLabels(node *corev1.Node) sets.String {
+	pools := sets.NewString()
+	eips, err := c.externalIPPoolLister.List(labels.Everything())
+	if err != nil {
+		klog.ErrorS(err, "Filter ExternalIPPools from nodeLabels failed")
+		return pools
+	}
+	for _, eip := range eips {
+		nodeSelector, _ := metav1.LabelSelectorAsSelector(&eip.Spec.NodeSelector)
+		if nodeSelector.Matches(labels.Set(node.GetLabels())) {
+			pools.Insert(eip.Name)
+		}
+	}
+	return pools
+}
+
+// Run will join other clusters if there are any Node in nodeLister,
```

Comment:
I don't understand this comment. What do you mean "other clusters", I thought there was a single cluster that all the Nodes are joining?
Maybe you mean:
> Run will join all the other K8s Nodes in a memberlist cluster

Reply:
Yes, you are right. If the local node has not joined any other nodes, we can look at it as a single cluster, I think.